hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -118,38 +118,35 @@ protected FsGetter fsGetter() {
    * Caching children filesystems. HADOOP-15565.
    */
   static class InnerCache {
-    private Map<Key, FileSystem> map = new HashMap<>();
+    private ConcurrentHashMap<Key, FileSystem> map = new ConcurrentHashMap<>();
     private FsGetter fsCreator;
-    private ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock();
 
     InnerCache(FsGetter fsCreator) {
       this.fsCreator = fsCreator;
     }
 
+    // computeIfAbsent() does not support a mapping function which throws IOException.
+    // Wrap fsCreator.getNewInstance() to not throw IOException and return null instead.
+    FileSystem getNewFileSystem(FsGetter fsCreator, URI uri, Configuration config) {
+      try {
+        return fsCreator.getNewInstance(uri, config);
+      } catch (IOException e) {
+        LOG.error("Failed to create new FileSystem instance for " + uri, e);
+        return null;
+      }
+    }
+
     FileSystem get(URI uri, Configuration config) throws IOException {
       Key key = new Key(uri);
-      FileSystem fs = null;
-      try {
-        rwLock.readLock().lock();
-        fs = map.get(key);
-        if (fs != null) {
-          return fs;
-        }
-      } finally {
-        rwLock.readLock().unlock();
+      if (map.contains(key)) {
```

Comment:
This should be unnecessary. The computed absent call returns existing mapped value if one exists. The whole point of the high level calls is that it is easy to make code thread safe when you just need 1 API call.

Reply:
Good point! Did not realize that. Fixed. thanks,

Reply:
This version is now much better than the original one. :) Thanks for your help.